repository/activity_repository/activity_repository.go

```diff
@@ -0,0 +1,40 @@
+package activity_repository
+
+import (
+	"encoding/json"
+	"gin/config"
+	"gin/models"
+	"gopkg.in/redis.v5"
+)
+
+func (rs RedisActivityRepository) GetUserActivity(username string) ([]string, error) {
+	return rs.client.LRange(username, 0, 2).Result()
+}
+
+func (rs RedisActivityRepository) SetUserActivity(username string, request models.UserRequest) error {
+	reqJson, err := json.Marshal(request)
+	if err != nil {
+		return err
+	}
+
+	cmd := rs.client.LPush(username, reqJson)
+	if err := cmd.Err(); err != nil {
+		return err
+	}
+
+	err = rs.client.LTrim(username, 0, 2).Err()
+	return err
+}
+
+type ActivityService interface {
+	GetUserActivity(username string) ([]string, error)
+	SetUserActivity(username string, request models.UserRequest) error
+}
+
+type RedisActivityRepository struct {
+	client *redis.Client
+}
+
+func GetActivityRepository() ActivityService {
+	return RedisActivityRepository{config.RedisClient}
```

Comment:
Don't use a global Redis client in the struct instance.
You can create the Redis client and pass it as an argument to the constructor, or create the client inside the constructor and `return nil, error` if it fails.
WDYT?

Reply:
I've actually copied this from the Go searcher. why not using the global one?

Reply:
You can have a global `RedisActivityRepository`, but not use a global Redis client to instantiate possibly many instances.
We want every `RedisActivityRepository` to have its own client and to operate autonomously.

Reply:
A good solution we use a lot in `go_searcher` is using `sync.Once`. On the first call to `NewRedisActivityProvider` the client will be set up and on the following calls, just returned.
lookup `sync.Once `in go_searcher and you will find plenty of examples...